README.md
# bedrock-session-http

[![Build Status](http://ci.digitalbazaar.com/buildStatus/icon?job=bedrock-session-http)](http://ci.digitalbazaar.com/job/bedrock-session-http)

Provides session information via a REST API

## Requirements

- npm v3+

## Usage
  - GET on `routes.session` will refresh a session & return session data.
  - DELETE on `routes.session` will delete a session. 

  GET requests will return a JSON object containing session data.
  This object should include the `session.ttl` and other data added by
  libraries that listen for the `bedrock-session-http.session.get` event.
  The ttl maybe used client side to determine when a session will expire.


CHANGELOG.md
# bedrock-session-http ChangeLog

## 5.0.0 - 2022-04-29

### Changed
- **BREAKING**: Update peer deps:
  - `@bedrock/core@6`
  - `@bedrock/express@8`.

## 4.0.1 - 2022-04-01

### Fixed
- Use `jsdoc-to-markdown@7`.

## 4.0.0 - 2022-04-01

### Changed
- **BREAKING**: Rename package to `@bedrock/session-http`.
- **BREAKING**: Convert to module (ESM).
- **BREAKING**: Remove default export.
- **BREAKING**: Require node 14.x.

### Removed
- **BREAKING**: Remove unused route config for `logout` route.

## 3.4.0 - 2022-03-27

### Changed
- Update peer deps:
  - `bedrock@4.5`
  - `bedrock-express@6.4`.
- Use `esm.js` to transpile internals from ESM to CommonJS. Should be
  a non-breaking change.

## 3.3.0 - 2022-03-08

### Changed
- Update peer dependencies.
- Update test suite to use more modern libraries.

## 3.2.1 - 2020-01-05

### Removed
- Remove deprecated GET-based logout security vulnerability.

## 3.2.0 - 2019-11-08

### Changed
- Update to latest bedrock events API.

## 3.1.0 - 2018-05-24

### Added
- Add DELETE endpoint for destroying sessions.

## 3.0.2 - 2016-09-22

### Changed
- Restructure test framework for CI.

## 3.0.1 - 2016-07-22

### Fixed
- Ensure status code is set before sending session.

## 3.0.0 - 2016-04-26

### Changed
- Logout no longer does a redirect to '/'.

## 2.0.2 - 2016-04-15

### Changed
- Update bedrock dependencies.

## 2.0.1 - 2016-03-24

### Changed
- Changed name of module from bedrock-session-rest to bedrock-session-http

## 2.0.0 - 2016-03-03

### Changed
- Update package dependencies for npm v3 compatibility.

## 1.0.0 - 2015-10-27

- See git history for changes.


test/test.js
/*!
 * Copyright (c) 2015-2022 Digital Bazaar, Inc. All rights reserved.
 */
import * as bedrock from '@bedrock/core';
import '@bedrock/express';
import '@bedrock/server';
import '@bedrock/session-http';
import {mock} from './mocha/mock.data.js';

bedrock.events.on('bedrock-session-http.session.get', (req, session) => {
  if(req.session.userId) {
    session.user = mock.user;
  }
});

bedrock.events.on('bedrock-express.configure.routes', app => {
  app.post('/login', (req, res) => {
    req.session.userId = mock.user.id;
    res.json(mock.user);
  });
});

import '@bedrock/test';
bedrock.start();


bedrock/test.js
/*
 * Copyright (c) 2015-2016 Digital Bazaar, Inc. All rights reserved.
 */
var bedrock = require('bedrock');
require('../lib/session');

var user = null;

bedrock.events.on('bedrock-session-http.session.get', function(req, session) {
  if(user) {
    session.id = user.id;
  }
});

bedrock.events.on('bedrock-express.configure.routes', function(app) {
  app.post('/login', function(req, res) {
    user = bedrock.config['session-http'].test.user;
    res.json(user);
  });

  app.get('/logout', function(req, res) {
    user = null;
    res.json({status: 'success'});
  });
});

bedrock.start();


test/test.config.js
/*!
 * Copyright (c) 2015-2022 Digital Bazaar, Inc. All rights reserved.
 */
import {config} from '@bedrock/core';
import {fileURLToPath} from 'node:url';
import path from 'node:path';
import '@bedrock/express';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

config.mocha.tests.push(path.join(__dirname, 'mocha'));

config.express.useSession = true;


lib/main.js
/*!
 * Copyright (c) 2015-2022 Digital Bazaar, Inc. All rights reserved.
 */
import * as bedrock from 'bedrock';
import {asyncHandler} from 'bedrock-express';

import './config.js';

// add routes
bedrock.events.on('bedrock-express.configure.routes', function(app) {
  const {routes} = bedrock.config['session-http'];

  app.get(routes.session, asyncHandler(async (req, res) => {
    // emit event with request and session and let listeners add session data
    const session = {};
    await bedrock.events.emit(
      'bedrock-session-http.session.get', req, session);
    res.status(200).json(session);
  }));

  app.delete(routes.session, (req, res, next) => {
    if(req.session) {
      return req.session.destroy(err => {
        if(err) {
          return next(err);
        }
        res.sendStatus(200);
      });
    }
    res.sendStatus(404);
  });
}); // end routes


lib/index.js
/*!
 * Copyright (c) 2015-2022 Digital Bazaar, Inc. All rights reserved.
 */
import * as bedrock from '@bedrock/core';
import {asyncHandler} from '@bedrock/express';

import './config.js';

// add routes
bedrock.events.on('bedrock-express.configure.routes', function(app) {
  const {routes} = bedrock.config['session-http'];

  app.get(routes.session, asyncHandler(async (req, res) => {
    // emit event with request and session and let listeners add session data
    const session = {};
    await bedrock.events.emit(
      'bedrock-session-http.session.get', req, session);
    res.status(200).json(session);
  }));

  app.delete(routes.session, (req, res, next) => {
    if(req.session) {
      return req.session.destroy(err => {
        if(err) {
          return next(err);
        }
        res.sendStatus(200);
      });
    }
    res.sendStatus(404);
  });
}); // end routes


lib/session.js
/*
 * Bedrock Session HTTP module.
 *
 * Copyright (c) 2015-2016 Digital Bazaar, Inc. All rights reserved.
 */
var bedrock = require('bedrock');
require('bedrock-express');

// module API
var api = {};
module.exports = api;

require('./config');

bedrock.events.on('bedrock.test.configure', function() {
  // load test config
  require('./test.config');
});

var routes = bedrock.config['session-http'].routes;

// add routes
bedrock.events.on('bedrock-express.configure.routes', function(app) {
  app.get(routes.session, function(req, res, next) {
    // emit event with request and session and let listeners add session data
    var session = {};
    bedrock.events.emit(
      'bedrock-session-http.session.get', req, session, function(err) {
      if(err) {
        return next(err);
      }
      res.status(200).json(session);
    });
  });

  app.get(routes.logout, function(req, res, next) {
    if(req.session) {
      return req.session.destroy(function(err) {
        if(err) {
          next(err);
        }
        res.sendStatus(200);
      });
    }
    res.sendStatus(404);
  });
}); // end routes
